temporal-sdk/src/main/java/io/temporal/common/SearchAttributes.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2022 Temporal Technologies, Inc. All Rights Reserved.
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.common;
+
+import java.util.*;
+import javax.annotation.Nonnull;
+
+/**
+ * Immutable collection of typed search attributes. Use {@link Builder} to create this collection.
+ */
+public final class SearchAttributes {
+  /** An empty search attribute collection. */
+  public static final SearchAttributes EMPTY = new SearchAttributes(Collections.emptySortedMap());
+
+  /** Create a new builder to create a search attribute collection. */
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  /**
+   * Create a new builder to create a search attribute collection, copying from an existing
+   * collection.
+   */
+  public static Builder newBuilder(SearchAttributes copyFrom) {
+    return new Builder(copyFrom);
+  }
+
+  /** Builder for creating a search attribute collection. */
+  public static class Builder {
+    private SortedMap<SearchAttributeKey<?>, Object> untypedValues;
+
+    private Builder() {
+      untypedValues = new TreeMap<>();
+    }
+
+    private Builder(SearchAttributes copyFrom) {
+      untypedValues = new TreeMap<>(copyFrom.untypedValues);
+    }
+
+    /** Set a search attribute key and typed value. */
+    public <T> Builder set(SearchAttributeKey<T> key, @Nonnull T value) {
+      Objects.requireNonNull(untypedValues, "Collection already built");
+      Objects.requireNonNull(value, "Value cannot be null");
+      // TODO(cretz): Prevent duplicate key name?
+      untypedValues.put(key, value);
```

Comment:
IMO override and mention it in the docstring. Mildly annoying to have to strip all-but-last if you want to map a series of changes to the final upsert, which is probably a common case.

Reply:
I concur, this jives with `put` (and probably don't need to even mention it since "set" implies it as opposed to "add")

Reply:
Removed TODO update doc